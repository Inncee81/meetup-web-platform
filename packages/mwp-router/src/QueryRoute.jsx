import React from 'react';
import PropTypes from 'prop-types';
import withRouter from 'react-router-dom/withRouter';

// determine whether the current querystring matches the specified param and
// optional paramVal
const getIsMatch = ({ location, param, value }) => {
	const params = new URLSearchParams(location.search);
	const paramVal = params.get(param);
	return value ? value === params.get(param) : params.has(param);
};

/*
 * This component conditionally renders its children based on the presence of
 * a particular querystring param (optionally with a specific value)
 * 
 * example:
 * 
 * ```
 * <QueryRoute param='foo' val='bar'>
 *   <div>This will display when '?foo=bar'</div>
 * <QueryRoute>
 * ```
 */
export class QueryRoute extends React.Component {
	constructor(props) {
		super(props);
		this.state = { isMatch: getIsMatch(props) };
	}
	componentWillReceiveProps(newProps) {
		if (newProps.location.search !== this.props.location.search) {
			this.setState(() => ({ isMatch: getIsMatch(newProps) }));
		}
	}
	render() {
		if (!this.state.isMatch) {
			return null;
		}
		return this.props.children;
	}
}

QueryRoute.propTypes = {
	location: PropTypes.object.isRequired, // provided by withRouter
	param: PropTypes.string.isRequired,
	value: PropTypes.string,
};
export default withRouter(QueryRoute);
