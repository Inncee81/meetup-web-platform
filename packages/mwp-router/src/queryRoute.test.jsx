import React from 'react';
import { shallow } from 'enzyme';
import { QueryRoute } from './QueryRoute';

const Dummy = () => <div>Hi</div>;

describe('QueryRoute', () => {
	it('renders child when querystring contains param, no value specified', () => {
		expect(
			shallow(
				<QueryRoute param="foo" location={{ search: '?foo' }}>
					<Dummy />
				</QueryRoute>
			)
		).toMatchSnapshot();
		expect(
			shallow(
				<QueryRoute param="foo" location={{ search: '?foo=bar' }}>
					<Dummy />
				</QueryRoute>
			)
		).toMatchSnapshot();
		expect(
			shallow(
				<QueryRoute param="foo" location={{ search: '?bar&foo' }}>
					<Dummy />
				</QueryRoute>
			)
		).toMatchSnapshot();
	});
	it('renders child when querystring contains param with specified value', () => {
		expect(
			shallow(
				<QueryRoute param="foo" value="bar" location={{ search: '?foo=bar' }}>
					<Dummy />
				</QueryRoute>
			)
		).toMatchSnapshot();
		expect(
			shallow(
				<QueryRoute
					param="foo"
					value="bar"
					location={{ search: '?qux&foo=bar' }}
				>
					<Dummy />
				</QueryRoute>
			)
		).toMatchSnapshot();
	});
	it('does not render anything when no matching param', () => {
		expect(
			shallow(
				<QueryRoute param="foo" location={{ search: '' }}>
					<Dummy />
				</QueryRoute>
			).isEmptyRender()
		).toBe(true);
		expect(
			shallow(
				<QueryRoute param="foo" location={{ search: '?bar' }}>
					<Dummy />
				</QueryRoute>
			).isEmptyRender()
		).toBe(true);
		expect(
			shallow(
				<QueryRoute param="foo" location={{ search: '?bar&food' }}>
					<Dummy />
				</QueryRoute>
			).isEmptyRender()
		).toBe(true);
	});
	it('does not render anything when specified param has non-matching value', () => {
		expect(
			shallow(
				<QueryRoute param="foo" value="bar" location={{ search: '?foo=qux' }}>
					<Dummy />
				</QueryRoute>
			).isEmptyRender()
		).toBe(true);
		expect(
			shallow(
				<QueryRoute
					param="foo"
					value="bar"
					location={{ search: '?foo=barn&food=bar' }}
				>
					<Dummy />
				</QueryRoute>
			).isEmptyRender()
		).toBe(true);
	});
});
