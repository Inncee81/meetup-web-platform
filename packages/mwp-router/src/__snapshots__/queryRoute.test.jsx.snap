// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryRoute renders child when querystring contains param with specified value 1`] = `<Dummy />`;

exports[`QueryRoute renders child when querystring contains param with specified value 2`] = `<Dummy />`;

exports[`QueryRoute renders child when querystring contains param, no value specified 1`] = `<Dummy />`;

exports[`QueryRoute renders child when querystring contains param, no value specified 2`] = `<Dummy />`;

exports[`QueryRoute renders child when querystring contains param, no value specified 3`] = `<Dummy />`;
